### PR TITLE
support new collections and opt-out cooling down

### PR DESCRIPTION
**Pull request by james97**
add cooling down gap for filtering events

utest

### CODE REVIEW

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

File: DEV.conf [hunk 1]
```diff
@@ -37,14 +37,15 @@
     "personalization_status_settings": {
         "db_host": "SOME_DB_HOST",
         "db_name": "SOME_DB_NAME",
-        "collection": "SOME_DB_COLLECTION"
+        "collection": "SOME_DB_COLLECTION",
+        "cooling_down_gap": 30
```
Comment (jinghe) [inline DEV.conf:41]: not sure what do you mean by `gap`

be more specific: `days_to_cool_down`, so people know it means the number of days.

File: recommendation/cf/personalization_status.py [hunk 2]
```diff
@@ -30,3 +54,27 @@ def fetch(self, user_id: str) -> Dict:
...
+        )  # Before all events started
+        all_records = list(
+            self._db_client.find({"user_id": user_id}).sort(TIMESTAMP_STR, -1)
+        )
+        for i in range(len(all_records) - 1):
+            if not all_records[i].get("personalization_status"):
+                continue
+            elif all_records[i].get(
+                "personalization_status"
+            ) and not all_records[i + 1].get("personalization_status"):
```
Comment (jinghe) [inline recommendation/cf/personalization_status.py:70]: `i+1` maybe over the boundary?

File: recommendation/cf/personalization_status.py [hunk 3]
```diff
@@ -30,3 +54,27 @@ def fetch(self, user_id: str) -> Dict:
             f"Get the latest personalization status {all_records[0]} for id: {user_id}"
         )
         return all_records[0]
+
+    def get_profile_event_start_date(self, user_id: str) -> datetime:
```
Comment (jinghe) [inline recommendation/cf/personalization_status.py:58]: describe what this method does in comments?

File: recommendation/cf/personalization_status.py [hunk 4]
```diff
@@ -1,19 +1,43 @@
...
+        """
         self._db_client = db_client
+        self._cooling_down_gap = cooling_down_gap
 
-    def fetch(self, user_id: str) -> Dict:
-        TIMESTAMP_STR = "timestamp"
-        all_records = list(self._db_client.find({"user_id": user_id}))
+    def fetch(
+        self, user_id: str, personalization_status: PersonalizationStatus = None
+    ) -> Dict:
```
Comment (jinghe) [inline recommendation/cf/personalization_status.py:33]: where did you call this extension?

File: recommendation/cf/personalization_status.py (hunk 2 above)
Comment (james97) [reply recommendation/cf/personalization_status.py:70]: it is
 
```
for i in range(len(all_records) - 1):
```

File: recommendation/cf/personalization_status.py (hunk 4 above)
Comment (james97) [reply recommendation/cf/personalization_status.py:33]: initially, it was used by `get_profile_event_start_date`, but I changed some logic so it is not used. I think it is still useful if we want to know the latest timestamp of a particular status

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

File: recommendation/cf/personalization_status.py [hunk 5]
```diff
@@ -30,3 +54,34 @@ def fetch(self, user_id: str) -> Dict:
             f"Get the latest personalization status {all_records[0]} for id: {user_id}"
         )
         return all_records[0]
+
+    def get_profile_event_start_date(self, user_id: str) -> datetime:
+        """
+        if a user disables personalization for more than `days_to_cool_down`, events before the
+        date that user enables personalization shouldn't be used anymore. So when we get user profile,
+        this method returns the start date of events we can use for personalization.
```
Comment (jinghe) [inline recommendation/cf/personalization_status.py:62]: put this line to the top of the comment. 

and then follows the description of some scenarios:
- disable/enable;
- no events found?
- in disable status?
- ...

File: recommendation/cf/personalization_status.py (hunk 2 above)
Comment (jinghe) [reply recommendation/cf/personalization_status.py:77]: what if it's empty or only has 1 record?

File: recommendation/cf/personalization_status.py (hunk 5 above)
Comment (james97) [reply recommendation/cf/personalization_status.py:62]: this method doesn't care about the current status of the user, nor there are events for the user or not. it just determines the event start date and UserProfileGenerator uses it to filter out events. Recommenders will get the latest user status.

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

File: recommendation/cf/personalization_status.py (hunk 2 above)
Comment (james97) [reply recommendation/cf/personalization_status.py:83]: nice catch. add an `if` statement above

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

File: recommendation/cf/personalization_status.py (hunk 4 above)
Comment (jinghe) [reply recommendation/cf/personalization_status.py:33]: so rename it to be more explicit. `fetch` is too fuzzy.

File: recommendation/cf/personalization_status.py (hunk 3 above)
Comment (jinghe) [reply recommendation/cf/personalization_status.py:58]: I think we need to return either `None` or a datetime in the future if the user's personalization status is not enabled, so we make sure the date is not misused.

**Review by jinghe — COMMENTED** [review]
c

File: recommendation/cf/personalization_status.py (hunk 3 above)
Comment (james97) [reply recommendation/cf/personalization_status.py:58]: I think that is out of the scope of this function. This status fetcher just returns the status. It doesn't need to check whether personalization is enabled or not and it doesn't care. Only recommenders or user profile generators care about that. If personalization is disabled, no matter what the date is, we shouldn't use any events

File: recommendation/cf/personalization_status.py (hunk 4 above)
Comment (james97) [reply recommendation/cf/personalization_status.py:33]: changed to `get_current_status`

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

File: recommendation/cf/personalization_status.py (hunk 3 above)
Comment (jinghe) [reply recommendation/cf/personalization_status.py:58]: so the date should be `None` logically.

**Comment by SiweiYang** [comment]
Something wrong, please check.

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

File: recommendation/cf/personalization_status.py (hunk 4 above)
Comment (jinghe) [reply recommendation/cf/personalization_status.py:33]: add comments please.

It's a bit confusing when you want to get the current status, but you ask the client to provide a `PersonalizationStatus` instance. explain what the parameter is used?

Also the return value is not a `PersonalizationStatus` but a Dict, need to explain what is expected in that Dict.

**Review by jinghe — COMMENTED** [review]
c

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

**Review by jinghe — APPROVED** [review]

**Comment by SiweiYang** [comment]
This build passes all tests. Good job!

**Pull request merged**